File: app/src/solver/BorderedWordNode.js
var quat = quat || {};
quat.solver = quat.solver || {};

quat.solver.BorderedWordNode = quat.solver.WordNode.extend({

	applyTheme: function(theme) {
        this._super(theme);

        this.setColor(this.default);

    },

    /**
     * Creates a WordNode with the given font size.
     * @param  {number} fontSize The size of the font to be initialized.
     * @param  {number} fontGap The distance between each letter.
     */
    ctor: function (fontSize, fontGap) {
        this._super(fontSize, fontGap);

        this.default = cc.color(255,255,255,64),
        this.selected = cc.color(255,255,255,153);

        var borders = [],
            fills = [],
        	size = fontSize * 1.25,
        	radius = fontSize * 0.24,
        	width = fontSize * 0.17;

        // Iterate through and create the rectangles
        for (var j = 0; j < 4; j++) {

            // Set up the border rectangle
            var border = new quat.RectRadius(size, size, radius, width, false),
            	bound = this.bounds[j];

            // Set them to be hidden
            border.x = bound.x + (bound.width / 2);
            border.y = bound.y + (bound.height / 2);
            border.zIndex = 1;

			border.setColor(cc.color(255,255,255,255));
            // Add it as a child to this layer
            this.addChild(border);

            borders.push(border);

            // Set up the fill rectangle
            var fill = new quat.RectRadius(size, size, radius, width, true);

            // Set them to be hidden
            fill.x = bound.x + (bound.width / 2);
            fill.y = bound.y + (bound.height / 2);
            fill.zIndex = 0;

            fill.setColor(this.default);
            // Add it as a child to this layer
            this.addChild(fill);

            fills.push(fill);
        }
        this.borders = borders;
        this.fills = fills;

        return true;
    },

    select: function(col) {
        this.fills[col].setColor(this.selected);
    },

    unselect: function() {
        for (var j = 0; j < 4; j++) {
            var rect = this.fills[j];
            rect.setColor(this.default);
        }
    },

    setOpacity: function(opacity) {
    	this._super(opacity);

    	for (var j = 0; j < 4; j++) {
            this.borders[j].setOpacity(opacity);
            this.fills[j].setOpacity(opacity);
        }
    },

    setColor: function(color) {
    	this._super(color);

    	for (var j = 0; j < 4; j++) {
            this.borders[j].setColor(color);
            this.fills[j].setColor(color);
        }
    }
});

File: cpp/Classes/solver/BorderedWordNode.cpp
#include "BorderedWordNode.h"

namespace QUAT {

bool BorderedWordNode::init() {
	// Init the super class
    if ( !WordNode::init() )
    {
        return false;
    }

    this->defaultColor = new cocos2d::Color4B(255,255,255,64);
    this->selectedColor = new cocos2d::Color4B(255,255,255,153);

    float size   = this->fontSize * 1.25,
    	  radius = this->fontSize * 0.24,
    	  width  = this->fontSize * 0.06;

    for (int i = 0; i < 4; i++) {
    	// Initialize the new RectRadius
    	this->borders[i] = RectRadius::create(size, size, radius, width, false);
    	auto border = this->borders[i];

    	// Grab the calculated bounds of the letter
    	auto bound = (*this->bounds)[i];

    	border->setPositionX(bound->origin.x + (bound->size.width / 2));
    	border->setPositionY(bound->origin.y + (bound->size.height / 2));

    	border->setColor(cocos2d::Color4B::WHITE);

    	this->addChild(border);

    	// Initialize the new RectRadius
    	this->fills[i] = RectRadius::create(size, size, radius, width, true);
    	auto fill = this->fills[i];

    	fill->setPositionX(bound->origin.x + (bound->size.width / 2));
    	fill->setPositionY(bound->origin.y + (bound->size.height / 2));

    	fill->setColor(*this->defaultColor);

    	this->addChild(fill);
    }

	return true;
}

void BorderedWordNode::select(int column) {
	this->fills[column]->setColor(*this->selectedColor);
}

void BorderedWordNode::unselect() {
	for (int i = 0; i < 4; i++) {
		auto fill = this->fills[i];
    	fill->setColor(*this->defaultColor);
    }
}

BorderedWordNode::BorderedWordNode(float fontSize, float fontGap) : WordNode(fontSize, fontGap) {
	
}

BorderedWordNode * BorderedWordNode::create(float fontSize, float fontGap)
{
    BorderedWordNode *pRet = new(std::nothrow) BorderedWordNode(fontSize, fontGap);
    if (pRet && pRet->init())
    {
        pRet->autorelease();
        return pRet;
    }
    else
    {
        delete pRet;
        pRet = NULL;
        return NULL;
    }
}

}